scipy/_lib/_uarray/_uarray_dispatch.cxx

```diff
@@ -140,30 +140,33 @@ Using these with PyObject_GetAttr is faster than PyObject_GetAttrString which
 has to create a new python string internally.
  */
 struct {
-  py_ref ua_convert;
-  py_ref ua_domain;
-  py_ref ua_function;
+  py_ref* ua_convert;
+  py_ref* ua_domain;
+  py_ref* ua_function;
 
   bool init() {
-    ua_convert = py_ref::steal(PyUnicode_InternFromString("__ua_convert__"));
-    if (!ua_convert)
+    static py_ref ua_convert_str = py_ref::steal(PyUnicode_InternFromString("__ua_convert__"));
```

Comment:
I may be missing something, but I don't understand how this actually fixes the issue. It looks like there's still a `py_ref` with global lifetime: `static py_ref ua_convert_str`.

Won't this get released when the shared library is unloaded (after the interpreter is destroyed)?

Reply:
Correct me if I'm wrong, is this missing an extra `Py_INCREF` call per string in order to leak them? Like in https://github.com/pytorch/pytorch/pull/130341/files#diff-9526caa8d74eda828570b8fcc06917713a198b0ed4fd622b49a8a6abf2f03f52R744

Reply:
It's not just a matter of a `Py_INCREF` to leak them, you don't want to call `Py_DECREF` from the destructors of variables with global lifetimes. Even if it doesn't trigger the destructor, the `Py_DECREF` after Python is finalized or without a valid thread state isn't safe, especially in debug builds or the free-threaded build.

You really do not want a `py_ref` with global lifetime. The `py_ref` here is only causing you problems. Just use a `PyObject *ua_convert`, etc.

There are a number of other problematic variables:
* `py_ref BackendNotImplementedError`
* `global_state_t global_domain_map` (struct contains py_refs)
* `thread_local global_state_t thread_local_domain_map` (struct contains py_refs)
* `thread_local local_state_t local_domain_map` (struct contains py_refs)

`thread_local` `py_ref` variables have a similar problem as global variables. The C++ destructor will be called when the thread doesn't hold the GIL (or have a valid thread state in the free-threaded build). That's not safe.

Reply:
Note that `TensorType` in the linked to PyTorch code is a `PyTypeObject*`, not a "smart pointer" like `py_ref` or pybind11's `py::object `.